File: minrotatearray.py
# Time Complexity : O(logn)
# Space Complexity : 0(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this :

class Solution:
    def findMin(self, nums: List[int]) -> int:
        # Error case
        if len(nums) == 0 or nums == None :
            return -1
        
        if len(nums) == 1:
            return nums[0]
        
        #1
        low = 0
        high = len(nums) - 1
        
        if nums[low] < nums[high]:
            return nums[low]
        #2
        while high >= low:
            mid = low + (high - low) //2
            
            if mid>0 and nums[mid] < nums[mid -1]:
                return nums[mid]
            
            elif nums[mid] > nums[high]: low = mid + 1
            else: high = mid - 1
        
        return -1

File: minrotatearray.java

// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
    public int findMin(int[] nums) {
       int n = nums.length;
       int left = 0, right = n - 1;
       while(left <= right){
           if(nums[left] < nums[right])
                return nums[left];
           int mid = left + (right - left)/2;
           if((mid==0 || nums[mid] < nums[mid-1]) && (mid == n - 1 || nums[mid]<nums[mid+1]))
            return nums[mid];
           else if(nums[mid] > nums[right])
               left = mid + 1;
            else
                right = mid - 1;
           }
           return nums[left];
        
       } 
    }